### PR TITLE
TauModel serialization to json 

### PR DESCRIPTION
**Pull request by megies**
Serialization of `TauModel` working on Py2, only missing the `.sMod` `SlownessModel` bit right now..

### CODE REVIEW

**Comment by QuLogic** [comment]
I'm not super-keen on text-serialization; I don't suppose there's any way to adapt this to NumPy's saving method?

**Comment by ThomasLecocq** [comment]
:trollface: why not using pandas ? :-) (no link, just kidding (and it's friday night))

**Comment by QuLogic** [comment]
I haven't read it all yet, but maybe something like `np.savez_compressed(fn, **data)` would work, since you've done the work of pulling out the relevant key/values.

**Comment by megies** [comment]
Yeah, I admit it's a bit ugly.. Considered `np.savez`, but I figured we can't use it because..
 - we always have arrays after deserializing.. even attributes that were originally a single float or a single bool (might not be very problematic)
 - `TauModel` has an attribute `.tauBranches` which is a 2D array of type `TauBranch`.. no chance for a simple `savez` here

**Comment by megies** [comment]
> :trollface: why not using pandas ? :-) (no link, just kidding (and it's friday night))

:trollface: sadly, no belgian beer for this sprint. but our brew has 95 by the [`THE BROS`](http://www.beeradvocate.com/beer/profile/820/2560/). cheers.

**Comment by QuLogic** [comment]
> we always have arrays after deserializing.. even attributes that were originally a single float or a single bool (might not be very problematic)

I _think_ it should be okay because they'd be 0-D, but might be trouble for any `isinstance` checks.

> TauModel has an attribute .tauBranches which is a 2D array of type TauBranch.. no chance for a simple savez here (show stopper)

@mbyt had some ideas in #958 for that; not sure how simple they are.

**Comment by megies** [comment]
> @mbyt had some ideas in #958 for that; not sure how simple they are.

I don't see anything in there about handling the object type of `TauBranch`.
It's a complex class and I don't see how it can be handled just by numpy standard routines.. or am I missing some magic here?

**Comment by QuLogic** [comment]
https://github.com/obspy/obspy/issues/958#issuecomment-74437364 - not _super_ simple, I'd guess.

**Comment by megies** [comment]
> #958 (comment) - not super simple, I'd guess.

Sorry, i see the structured array part but I still son't see the handling arrays of arbitrary object types..

**Comment by QuLogic** [comment]
Ah, you'd basically have to pivot them (obligatory pandas reference?) Go from array of `TauBranch.x` (single float) to `np.array` (for all branches) and from `TauBranch.a` (array) to 2D `np.array` (for all branches).

File: obspy/taup/slowness_model.py [hunk 1]
```diff
@@ -634,7 +637,7 @@ def depthInFluid(self, depth):
             otherwise.
         :rtype: :class:`~numpy.ndarray` (dtype = :class:`bool`)
         """
-        ret = np.zeros(shape=depth.shape, dtype=np.bool_)
+        ret = False
```
Comment (QuLogic) [inline obspy/taup/slowness_model.py]: I'm not sure about this?

**Comment by QuLogic** [comment]
Hmm, ~430K per model (vs 256K via pickle), but we don't need one for each Python version, so I guess that's a net win?

**Comment by megies** [comment]
> Hmm, ~430K per model (vs 256K via pickle), but we don't need one for each Python version, so I guess that's a net win?

Plus we can still gzip it, which makes it ~180K per model. But there's still a strange error, have to find out what is wrong with the de/serialized models..

**Comment by megies** [comment]
Hmm.. Actually I've been investigating the numpy way for serialization a bit now. I think I got a way layed out that should work pretty well.

Here's an example snippet I played with that shows how to serialize an object `A` that has an 2D array attribute of class `B` which c be represented with a structured array:

```python
import random
import numpy as np


class A(object):
    def __init__(self, seed=None):
        if seed is None:
            return
        np.random.seed(seed)
        self.a = np.random.randint(0, 100, 1)
        self.b = np.random.randint(0, 100, 10)
        my_bs = [B(np.random.randint(0, 100, 1))
                 for x in np.random.randint(0, 100, 10)]
        self.my_bs = np.array(my_bs).reshape((2, 5))

    def __str__(self):
        info = ", ".join(map(str, [self.a, self.b]))
        info += "\n" + "=" * 40 + "\n"
        info += "\n".join([str(b) for b in self.my_bs.flatten()])
        return info

    def serialize(self):
        i, j = self.my_bs.shape
        bs = np.vstack([np.hstack([self.my_bs[i_][j_]._to_array()
                                   for j_ in range(j)])
                        for i_ in range(i)])
        np.savez("/tmp/pickling/a.npz", a=self.a, b=self.b, my_bs=bs)

    @staticmethod
    def deserialize():
        arr = np.load("/tmp/pickling/a.npz")
        a = A()
        a.a = arr['a']
        a.b = arr['b']
        i, j = arr['my_bs'].shape
        a.my_bs = np.array([[B._from_array(arr['my_bs'][i_][j_])
                             for j_ in range(j)]
                            for i_ in range(i)])
        return a


class B(object):
    def __init__(self, seed=None):
        if seed is None:
            return
        np.random.seed(seed)
        self.x = np.random.randint(0, 100, 1)
        self.y = np.random.randint(0, 100, 10)

    def __str__(self):
        return ", ".join(map(str, [self.x, self.y]))

    def _to_array(self):
        arr = np.empty(shape=(), dtype=[('x', 'i4'),
                                        ('y', 'i4', self.y.shape)])
        arr['x'] = self.x
        arr['y'] = self.y
        return arr

    @staticmethod
    def _from_array(arr):
        b = B()
        b.x = arr['x']
        b.y = arr['y']
        return b
#
#    def serialize(self):
#        np.save("/tmp/pickling/b.npy", self._to_array())
#
#    @staticmethod
#    def deserialize():
#        return B._from_array(np.load("/tmp/pickling/b.npy"))


if __name__ == "__main__":
    b = B(random.randint(1, 1000))
    print b
    print
    a = A(random.randint(1, 1000))
    print a
    print
    a.serialize()
    a_ = A.deserialize()
    print a_
```

So I think I'll serialize `TauModel` using `np.savez`, the above 2D array of `TauBranch`es was what worried me the most before..

**Comment by mbyt** [comment]
Great example, makes sense to me. Even easier to understand the the `split(').reshape()` magic :).

> But there's still a strange error, have to find out what is wrong with the de/serialized models..

Does this resolve this strange error of the 4 extra values?

**Comment by megies** [comment]
Closing in favor of #984.

**Pull request closed without merging**